cpp/src/fil/infer.cu

```diff
@@ -310,6 +310,7 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES>
 
     __syncthreads();  // free up per_class_margin[]
     write_best_class_in_block(best, blockDim.x, out, num_rows);
+    __syncthreads();  // free per_class_margin[] for next row set
```

Comment:
even though it's unnecessary in case `blocks_per_sm == 0`, it doesn't affect timing even in the most sensitive cases for the above algo. This allows not to leak another implementation detail to `__global__ void infer_k`

Reply:
This is fine.